File: bun/src/services/realtime/executor/socket.executor.ts
import { NotificationType } from '../notification'
import { INotificationExecutor } from './notification.executor'
import { ISocketPublisher } from '../../socket/socket.publisher'
import { Notification } from '../notification.processor'
import Logger from '../../../config/logger'
const logger = Logger(import.meta.file)

export class SocketNotificationExecutor implements INotificationExecutor {
  publishers: ISocketPublisher[]
  constructor(pubsub: ISocketPublisher[]) {
    this.publishers = pubsub
  }

  async execute<T>(type: NotificationType, notification: Notification<T>): Promise<void> {
    logger.info(`SocketNotificationExecutor: ${type} for id: ${notification.id}`)
    const channel = `${type}:${notification.id}`
    const promises = this.publishers.map((publisher) => publisher.publish(channel, notification.message))
    await Promise.all(promises)
  }
}

File: go/internal/processors/executors/notification/socket_executor.go
package executors

import (
	"github.com/marlpativ/healthz-curler/internal/models"
	"github.com/marlpativ/healthz-curler/internal/processors/executors"
	"github.com/marlpativ/healthz-curler/internal/socket"
)

type socketNotificationExecutor struct {
	publishers []socket.SocketPublisher
}

func NewSocketNotificationExecutor(publishers []socket.SocketPublisher) executors.NotificationExecutor {
	return &socketNotificationExecutor{
		publishers: publishers,
	}
}

func (s *socketNotificationExecutor) Execute(notificationType string, notification models.Notification) {
	channel := notificationType + ":" + notification.Id
	for _, publisher := range s.publishers {
		publisher.Publish(channel, notification)
	}
}
